### DISCUSSION TITLE
Solar Trailer Monitoring and Control via Grott(server) and friends?

### DISCUSSION THREAD

**Discussion by fhteagle** [johanmeijer/grott#404] [Q&A]
Greetings all, and thanks for this great project that I have been testing as a data source for Home Assistant for about a month now with good results. But I am thinking ahead to the ultimate goal, which is to have my 2 Growatt units + batteries + panels mounted on a trailer to have mobile power at various places around my property throughout the year. This means the trailer will sometimes be in range of my home Wifi, but sometimes will be too far away to report to the MQTT/InfluxDB Docker containers hosted at the house. So I am looking for a solution to make the trailer stand on its own for monitoring and control sometimes, but also report up to the rest of the network stack once it comes home. I would appreciate any suggestions, feedback, recipes, etc that you think might help me to plan this.

Hardware I plan to use (mostly on hand already):
- 2 Growatt SPF-3000 TL, usually run in split phase to make 240V, but might be rarely reconfigured to be paralleled in single phase 120V, then back to 240V again later. I assume the reconfiguration from 240V to 120V would involve manual wiring changes and/or inputs to the front control panel of the Growatts, but if I could make this a single electronic command, DPDT switch control, or momentary push button toggle action, that would be even better.
- Daly BMS (will occupy the RS485 slot or slots)
- LFP batteries
- My single ShineWifi-F datalogger module only monitors and reports on the unit it is connected to, not its paired partner unit. So I am thinking I would need to DIY up the ability to read from both units via USB. Relevant projects to this might be https://github.com/Lyve1981/GrowattUsbModbus and https://github.com/otti/Growatt_ShineWiFi-S .
- I have some old Raspberry Pis (2B 's, 3B(+), and a zero W) that I could re-purpose to be the hardware controller for all this
- USB controlled relays and automotive grade physical contactor relays
- DC battery heating pads

Desired features:
- When parked at home / within wifi range (mostly done/working already): 
-- joins the my home wifi network
-- passes operational data immediately to InfluxDB

- When away from home / out of wifi range: 
-- automatically puts up own wifi network (probably via hostapd triggered by systemd service, out of scope for Grott, and something I know how to do no problem) 
-- monitor/control Growatt operation and battery statistics via web portal on my Android phone(s)
-- logs data history locally on the controller device, but then data dumps the history to InfluxDB when the trailer gets back home days or weeks later

- Whether at home or away: 
-- I would love the ability to bring the AC output out of standby timeout by wifi command (from HomeAssistant when at home), from the web portal on the smartphone (when away from home). I understand this is a future feature planned for the Grott ecosystem at some point in the future, yesno?
-- trigger extra battery heating via HomeAssistant, web portal, and/or momentary push button or SPDT switch

Possible / practical? Not going to happen? Better ways to do this than what I have found and thought up so far? Let me know what you think. 

Thanks.

**Comment by johanmeijer** [comment]
Ok. That is real thinking out of the box. The challenge is that you want to able to communicate with the inverter/datalogger and would be able to write to a the influxdb in a later stage. 

Grott is a proxy and does not store data. So if the data is sent to growatt (or grottserver) and is processed by Grott then it is gone (from Grott point of view). 

A simple solution might be to create a extension that writes the data record to a file and then write a simple python script to read the file and process the data (the python script can execute grottdata.py, that process the data and write to influx). 
The challange for now is that Grott only supports 1 extension and that is probably alread in use for the HA extension. 
An other challenge (and I am writing and brainstorming at the same time) is that you only want to store records if inlfuxdb is not available (e.g. when you are not at home). 

When yo do not have an internet connection at all it should not be a problem. The datalogger will store the data and send it at the moment the connection is established.

**Reply by fhteagle** [reply]
Thanks for the reply and ideas.

I think my best case is to replace the OEM datalogger with a Raspberry Pi type device that can receive and sum the data from both units at once. Hence the first of the other projects I linked that I think can get me the reading in and parsing MODBUS, if I am reading their descriptions correctly.

I will look into the Grott extension idea. I do not need the HA extension loaded, I have my HA instance(s) able to ingest current data via MQTT thus far.
